Fix progressBar endDuration when using `stages`

ppcano opened the pull request:
closes #347

ppcano commented:
`master` rebased.

liclac commented:
Wrong place to fix this. Never duplicate information, that's a hack that *never* ends well. Instead change the `Right` function for the progress bar to check both what `GetEndDuration()` and `GetStages()` say, and use whichever one is valid (for `GetStages()`, `len(stages) > 0`) and lower.

ppcano commented:
@liclac Thanks.

Now, I have updated only the `Right` function.  The first attempt also updated the `duration` value of the test summary. The new change does not update the `duration` ( - ), but I am not sure if this is desired. 


![screen shot 2017-11-29 at 13 54 10](https://user-images.githubusercontent.com/825430/33376147-c2f34bd8-d50c-11e7-850b-b527aae58a7d.png)

The intention of the PR was to clarify the duration when using `stages`; IMO, updating the `progressbar` is enough to clarify it. We could evaluate later if more changes are required.

liclac commented on cmd/run.go:
```diff
@@ -268,7 +268,13 @@ a commandline interface for interacting with it.`,
 				precision := 100 * time.Millisecond
 				atT := engine.Executor.GetTime()
-				if endT := engine.Executor.GetEndTime(); endT.Valid {
+
```
Drop this newline.

liclac commented on cmd/run.go:
```diff
@@ -268,7 +268,13 @@ a commandline interface for interacting with it.`,
-				if endT := engine.Executor.GetEndTime(); endT.Valid {
+
+				endT := engine.Executor.GetEndTime()
+				if stages := engine.Executor.GetStages(); len(stages) > 0 {
```
Right, I forgot `SumStages()` returns a `NullDuration`, just make this an `endT := SumStages(...); endT.Valid`.

liclac commented on cmd/run.go:
```diff
@@ -268,7 +268,13 @@ a commandline interface for interacting with it.`,
+
+				endT := engine.Executor.GetEndTime()
+				if stages := engine.Executor.GetStages(); len(stages) > 0 {
+					endT = local.SumStages(stages)
```
If this is gonna be used outside of `local`, move it to `lib`.

liclac commented:
Just update the progress bar with the result of `SumStages()` then.

https://github.com/ppcano/k6/blob/926bb0089d71d82cd5ed398ed6e0b0ca34134842/cmd/run.go#L322

ppcano commented:
@liclac I have added your changes.

I am not sure how to implement if/else in a more idiomatic way:

```
endT := lib.SumStages(engine.Executor.GetStages())
if !endT.Valid {
   endT = engine.Executor.GetEndTime()
}
if endT.Valid {
     prog = float64(engine.Executor.GetTime()) / float64(endT.Duration)
}

```
```
endT := lib.SumStages(engine.Executor.GetStages())
if !endT.Valid {
    endT = engine.Executor.GetEndTime()
}
if endT.Valid {
   ...					
}
```

Feel free to suggest. Thanks.

liclac commented:
That's fine.

But you also just switched the priority (stages > end). The correct behaviour would be to get both times, then use whichever is valid and lower.

```go
endT := GetEndDuration()
stageEndT := SumStages()
if !endT.Valid || endT.Duration < stageEndT.Duration {
  endT = stageEndT
}
```

ppcano commented:
@liclac 

> The correct behaviour would be to get both times, then use whichever is valid and lower.

Done at 5785f20cd465e850bdc984d0f1f0927bc51a87df

```
if !endT.Valid || endT.Duration > stageEndT.Duration {
  endT = stageEndT
}
```

Let me know if you want me to `squash` all the commits or there is something else to include.

liclac commented on cmd/run.go:
```diff
@@ -273,7 +273,13 @@ a commandline interface for interacting with it.`,
-				if endT := engine.Executor.GetEndTime(); endT.Valid {
+				stagesEndT := lib.SumStages(engine.Executor.GetStages())
+				endT := engine.Executor.GetEndTime()
+				//get both times and use whichever is valid and lower
```
This comment is kind of unnecessary. If you feel it needs to be here, capitalise the G and put a space after the "//".

ppcano commented:
@liclac Removed the comment. Done, Thanks.
